builder/builder-next/exporter/wrapper.go

```diff
@@ -11,7 +11,7 @@ import (
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
-type ImageExportedByBuildkit = func(ctx context.Context, id string, desc ocispec.Descriptor) error
+type ImageExportedByBuildkit = func(ctx context.Context, id string, desc ocispec.Descriptor)
```

Comment:
Even starting to wonder if we need this type at all; it's used as type on some structs and as arguments, so things should already fail if the signature changes. Admitted, it will slightly more verbose to define it inline

```patch
diff --git a/builder/builder-next/builder.go b/builder/builder-next/builder.go
index 3ea8b90346..a13113eae2 100644
--- a/builder/builder-next/builder.go
+++ b/builder/builder-next/builder.go
@@ -34,6 +34,7 @@ import (
 	"github.com/moby/buildkit/session"
 	"github.com/moby/buildkit/util/entitlements"
 	"github.com/moby/buildkit/util/tracing"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/pkg/errors"
 	"golang.org/x/sync/errgroup"
 	"google.golang.org/grpc"
@@ -94,7 +95,7 @@ type Opt struct {
 	Snapshotter           string
 	ContainerdAddress     string
 	ContainerdNamespace   string
-	ImageExportedCallback exporter.ImageExportedByBuildkit
+	ImageExportedCallback func(ctx context.Context, id string, desc ocispec.Descriptor)
 }
 
 // Builder can build using BuildKit backend
diff --git a/builder/builder-next/exporter/mobyexporter/export.go b/builder/builder-next/exporter/mobyexporter/export.go
index dd78e3c165..7f705a75e9 100644
--- a/builder/builder-next/exporter/mobyexporter/export.go
+++ b/builder/builder-next/exporter/mobyexporter/export.go
@@ -10,7 +10,6 @@ import (
 	"github.com/containerd/containerd/leases"
 	"github.com/containerd/log"
 	distref "github.com/distribution/reference"
-	builderexporter "github.com/docker/docker/builder/builder-next/exporter"
 	"github.com/docker/docker/image"
 	"github.com/docker/docker/layer"
 	"github.com/moby/buildkit/exporter"
@@ -38,7 +37,7 @@ type Opt struct {
 	ImageTagger           ImageTagger
 	ContentStore          content.Store
 	LeaseManager          leases.Manager
-	ImageExportedCallback builderexporter.ImageExportedByBuildkit
+	ImageExportedCallback func(ctx context.Context, id string, desc ocispec.Descriptor)
 }
 
 type imageExporter struct {
diff --git a/builder/builder-next/exporter/wrapper.go b/builder/builder-next/exporter/wrapper.go
index 1676466ab9..3ac5d18735 100644
--- a/builder/builder-next/exporter/wrapper.go
+++ b/builder/builder-next/exporter/wrapper.go
@@ -11,18 +11,16 @@ import (
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
-type ImageExportedByBuildkit = func(ctx context.Context, id string, desc ocispec.Descriptor) error
-
 // Wraps the containerimage exporter's Resolve method to apply moby-specific
 // overrides to the exporter attributes.
 type imageExporterMobyWrapper struct {
 	exp      exporter.Exporter
-	callback ImageExportedByBuildkit
+	callback func(ctx context.Context, id string, desc ocispec.Descriptor)
 }
 
 // NewWrapper returns an exporter wrapper that applies moby specific attributes
 // and hooks the export process.
-func NewWrapper(exp exporter.Exporter, callback ImageExportedByBuildkit) (exporter.Exporter, error) {
+func NewWrapper(exp exporter.Exporter, callback func(ctx context.Context, id string, desc ocispec.Descriptor)) (exporter.Exporter, error) {
 	return &imageExporterMobyWrapper{exp: exp, callback: callback}, nil
 }
 
@@ -51,7 +49,7 @@ func (e *imageExporterMobyWrapper) Resolve(ctx context.Context, id int, exporter
 
 type imageExporterInstanceWrapper struct {
 	exporter.ExporterInstance
-	callback ImageExportedByBuildkit
+	callback func(ctx context.Context, id string, desc ocispec.Descriptor)
 }
 
 func (i *imageExporterInstanceWrapper) Export(ctx context.Context, src *exporter.Source, inlineCache exptypes.InlineCache, sessionID string) (map[string]string, exporter.DescriptorReference, error) {
diff --git a/builder/builder-next/worker/containerdworker.go b/builder/builder-next/worker/containerdworker.go
index 2f640cca69..17bfbdc0e4 100644
--- a/builder/builder-next/worker/containerdworker.go
+++ b/builder/builder-next/worker/containerdworker.go
@@ -8,16 +8,17 @@ import (
 	"github.com/moby/buildkit/exporter"
 	"github.com/moby/buildkit/session"
 	"github.com/moby/buildkit/worker/base"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
 // ContainerdWorker is a local worker instance with dedicated snapshotter, cache, and so on.
 type ContainerdWorker struct {
 	*base.Worker
-	callback mobyexporter.ImageExportedByBuildkit
+	callback func(ctx context.Context, id string, desc ocispec.Descriptor)
 }
 
 // NewContainerdWorker instantiates a local worker.
-func NewContainerdWorker(ctx context.Context, wo base.WorkerOpt, callback mobyexporter.ImageExportedByBuildkit) (*ContainerdWorker, error) {
+func NewContainerdWorker(ctx context.Context, wo base.WorkerOpt, callback func(ctx context.Context, id string, desc ocispec.Descriptor)) (*ContainerdWorker, error) {
 	bw, err := base.NewWorker(ctx, wo)
 	if err != nil {
 		return nil, err
```

Reply:
☝️ I went through the changes per-commit, so some of this may have changed with my follow-up comments; I haven't checked what it would look like if the `event.Action` would be added to the signature (perhaps in that case, it becomes really verbose, but not sure)

Reply:
Inlined